tokio/src/task/scope/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,628 @@
+            data,
+            _pin: PhantomPinned,
+        }
+    }
+}
+
+impl<T> Deref for ListNode<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.data
+    }
+}
+
+impl<T> DerefMut for ListNode<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+}
+
+/// An intrusive linked list of nodes, where each node carries associated data
+/// of type `T`.
+#[derive(Debug)]
+pub(crate) struct LinkedList<T> {
+    head: *mut ListNode<T>,
+    tail: *mut ListNode<T>,
+}
+
+impl<T> LinkedList<T> {
+    /// Creates an empty linked list
+    pub(crate) fn new() -> Self {
+        LinkedList::<T> {
+            head: null_mut(),
+            tail: null_mut(),
+        }
+    }
+
+    /// Consumes the list and creates an iterator over the linked list.
+    /// This function is only safe as long as all pointers which are stored inside
+    /// the linked list are valid.
+    #[allow(dead_code)]
+    pub(crate) unsafe fn into_iter(self) -> LinkedListIterator<T> {
+        LinkedListIterator { current: self.head }
+    }
+
+    /// Consumes the list and creates an iterator over the linked list in reverse
+    /// direction.
+    /// This function is only safe as long as all pointers which are stored inside
+    /// the linked list are valid.
+    pub(crate) unsafe fn into_reverse_iter(self) -> LinkedListReverseIterator<T> {
+        LinkedListReverseIterator { current: self.tail }
+    }
+
+    /// Adds an item to the front of the linked list.
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    pub(crate) unsafe fn add_front(&mut self, item: *mut ListNode<T>) {
+        assert!(!item.is_null(), "Can not add null pointers");
+        (*item).next = self.head;
+        (*item).prev = null_mut();
+        if !self.head.is_null() {
+            (*self.head).prev = item;
+        }
+        self.head = item;
+        if self.tail.is_null() {
+            self.tail = item;
+        }
+    }
+
+    /// Returns the first item in the linked list without removing it from the list
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    /// The returned pointer is only guaranteed to be valid as long as the list
+    /// is not mutated
+    #[allow(dead_code)]
+    pub(crate) fn peek_first(&self) -> *mut ListNode<T> {
+        self.head
+    }
+
+    /// Returns the last item in the linked list without removing it from the list
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    /// The returned pointer is only guaranteed to be valid as long as the list
+    /// is not mutated
+    #[allow(dead_code)]
+    pub(crate) fn peek_last(&self) -> *mut ListNode<T> {
+        self.tail
+    }
+
+    /// Removes the last item from the linked list and returns it
+    #[allow(dead_code)]
+    pub(crate) unsafe fn remove_last(&mut self) -> *mut ListNode<T> {
```

Comment:
This function doesn't look like it has to be `unsafe`. It has exclusive access, and all invariants are internally checked as far as I can see / understand. Otherwise, please document safety :D.